### PR TITLE
Optimize resource list query

### LINKED ISSUE #3374: Query for resources list used by web-console is slow
The query that is used for the web-console to list all versions is quite slow. On our internal mgmt orchestrator with thousands of resources it takes more than 500ms of which 90% is spend on sorting the data.

The concerning query for the mgmt environment is:

```sql
EXPLAIN (ANALYZE, COSTS, VERBOSE, BUFFERS, FORMAT JSON) SELECT * FROM (
        SELECT DISTINCT ON (r.resource_id) r.resource_id, r.attributes, r.resource_version_id, r.resource_type,
            r.agent, r.resource_id_value, r.environment,
            (CASE WHEN
                (SELECT r.model < MAX(public.configurationmodel.version)
                    FROM public.configurationmodel
                    WHERE public.configurationmodel.released=TRUE
                            AND environment='f81be80b-8582-499f-b002-7cdd2f2df98a')
                THEN 'orphaned'
            ELSE r.status::text END) as status
        FROM resource as r
        JOIN public.configurationmodel as cm ON r.model=cm.version AND r.environment=cm.environment
        WHERE cm.released=TRUE AND r.environment='f81be80b-8582-499f-b002-7cdd2f2df98a'
    )
as res
    ORDER BY resource_type ASC, resource_version_id ASC LIMIT 20;
```

This output is json and can be visualized with online tools. With `EXPLAIN (ANALYZE, COSTS, VERBOSE)` it is useful on the command line.

### PR DESCRIPTION
**Pull request by andraskvr**
# Description

The optimized query is based on https://stackoverflow.com/a/25536748, 1.a variant. I also added this link to the source

With the changes the average execution time on the same data (and database version) goes from the 500-600 ms range (https://explain.dalibo.com/plan/uWK#plan) to the 100-200 ms range (https://explain.dalibo.com/plan/yS8). 
Note: These results are based on the query plan analyzer and not on extensive performance testing.

closes #3374 

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [x] Attached issue to pull request
- [x] Changelog entry
- [x] Type annotations are present
- [x] Code is clear and sufficiently documented
- [x] No (preventable) type errors (check using make mypy or make mypy-diff)
- [x] Sufficient test cases (reproduces the bug/tests the requested feature)
- [x] Correct, in line with design
- [x] ~End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )~

# Reviewer Checklist:

- [ ] Sufficient test cases (reproduces the bug/tests the requested feature)
- [ ] Code is clear and sufficiently documented
- [ ] Correct, in line with design

### CODE REVIEW

**Review by wouterdb — COMMENTED** [review]
1. How much did we improve performance of the paged and non-paged versions of the query? 
2. I must conclude I don't understand this at all, it could do with some more comments as to WHY this is faster.

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

File: tests/server/test_resource_list.py [hunk 1]
```diff
@@ -138,13 +147,13 @@ async def env_with_resources(server, client):
...
     # Add multiple versions of model, with 2 of them released
-    for i in range(1, 4):
+    for i in range(1, 5):
         cm = data.ConfigurationModel(
             environment=env.id,
             version=i,
             date=datetime.now(),
             total=1,
-            released=i != 1,
+            released=i != 1 and i != 5,
```
Comment (arnaudsjs) [inline tests/server/test_resource_list.py:156]: `i` cannot be `5`

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

File: tests/server/test_resource_list.py [hunk 2]
```diff
@@ -162,12 +171,12 @@ async def create_resource(
             )
             await res.insert()
 
-    await create_resource("agent1", "/etc/file1", "std::File", ResourceState.available, [1, 2, 3])
+    await create_resource("agent1", "/etc/file1", "std::File", ResourceState.available, [1, 2, 3, 4])
```
Comment (arnaudsjs) [inline tests/server/test_resource_list.py:174]: This fixture is updated, but the fixture is not used by the test case that was updated?

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

**Comment by andraskvr** [comment]
> 1. How much did we improve performance of the paged and non-paged versions of the query?

In this query we need the resources with their maximal version numbers, and all of the approaches I've seen rely on the ordering (`resource_id, model DESC`) to build this up, in one way or the other, and after that we can do the paging on it with different orders, which has the disadvantage that this intermediate result set is built up more or less every time, regardless of paging. In practice the difference (based on a few experiments) is that a non-paged query takes about 190-200 ms, and a paged one takes 170-185 ms.

(The query would be a lot simpler / faster without the orphaned resources I think, so if we were only interested in resources in the latest version, not in resources with their own latest version)

>     2. I must conclude I don't understand this at all, it could do with some more comments as to WHY this is faster.

I added some more comments

File: tests/server/test_resource_list.py (hunk 2 above)
Comment (andraskvr) [reply tests/server/test_resource_list.py]: Good catch, this really didn't need to be changed

File: tests/server/test_resource_list.py [hunk 3]
```diff
@@ -74,13 +74,13 @@ async def test_has_only_one_version_from_resource(server, client):
     await env.insert()
 
     # Add multiple versions of model, with 2 of them released
-    for i in range(1, 4):
+    for i in range(1, 5):
```
[the remark this answers is not in the document]
Comment (andraskvr) [reply tests/server/test_resource_list.py:77]: I wanted to add some extra assertions to make sure it behaves exactly like the previous query.
But conceptually you are right, this is only a performance optimization PR, so I can revert the changes to the test file

File: changelogs/unreleased/3374-optimize-resource-list-query.yml [hunk 4]
```diff
@@ -0,0 +1,3 @@
+description: Optimize resource list query
+change-type: patch
+destination-branches: [master, iso4, iso5]
```
[the remark this answers is not in the document]
Comment (andraskvr) [reply changelogs/unreleased/3374-optimize-resource-list-query.yml:3]: We don't have a `performance` or similar section, so I added it to the `bugfix` section. The other alternative I think is `other-note`, what do you think?

[the remark this answers is not in the document]
Comment (arnaudsjs) [reply changelogs/unreleased/3374-optimize-resource-list-query.yml:3]: I don't have a strong preference here. `bugfix` is fine for me.

**Review by arnaudsjs — APPROVED** [review]

**Pull request closed without merging**